app.go

```diff
@@ -720,6 +742,63 @@ func (a *Application) processTx(txBytes []byte, isCheckTx bool) (TxHandlerResult
 	return r, nil
 }
 
+func (a *Application) deliverTx2(storeTx store.KVStoreTx, txBytes []byte) abci.ResponseDeliverTx {
+	state := NewStoreState(
+		context.Background(),
+		storeTx,
+		a.curBlockHeader,
+		a.curBlockHash,
+		a.GetValidatorSet,
+	).WithOnChainConfig(a.config)
+
+	receiptHandler := a.ReceiptHandlerProvider.Store()
+	defer receiptHandler.DiscardCurrentReceipt()
+	defer a.EventHandler.Rollback()
+
+	r, txErr := a.TxHandler.ProcessTx(state, txBytes, false)
+
+	// Store the receipt even if the tx itself failed
+	var receiptTxHash []byte
+	if a.ReceiptHandlerProvider.Reader().GetCurrentReceipt() != nil {
+		receiptTxHash = a.ReceiptHandlerProvider.Reader().GetCurrentReceipt().TxHash
+		txHash := ttypes.Tx(txBytes).Hash()
+		// If a receipt was generated for an EVM tx add a link between the TM tx hash and the EVM tx hash
+		// so that we can use it to lookup relevant events using the TM tx hash.
+		if !bytes.Equal(txHash, receiptTxHash) {
```

Comment:
We are creating an unnecessary link between TM Tx Hash and Receipt Tx Hash for EVM txs.
Receipt Tx hash is used as tx hash for EVM tx. So there is no need to create a link between TM tx has and TM Tx hash of a EVM transaction.

Reply:
For GO tx calling EVM, this is still needed.

Reply:
Yes the link is somewhat redundant for `TM(EVM)` txs, only really need it for `TM(GO->EVM)` txs, but this is meant to be an index. And for `TM(GO->EVM)` txs it means we don't have to hit `state.db` or `tx_index.db` to lookup the tx result to figure out what the EVM tx hash is. Ultimately we should probably just fix the Call/DeployTxHandler so they always store the tx type & EVM tx hash (if any EVM calls were made) in the `TxHandlerResult`.

Reply:
Got it